PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/algorithms/CompareSubgraphs.java

```diff
@@ -0,0 +1,257 @@
+      OriginalGraph bottomGraph) {
+    // todo: decolapse bubble from the graphs (possibly remember bubbles here for easy recolapsing)
+    OverlapThread overlapThread = new OverlapThread(topGraph, bottomGraph);
+    SubGraphOrderer topGraphOrderer = new SubGraphOrderer(mainGraphOrder, topGraph, overlapThread);
+    SubGraphOrderer bottomGraphOrderer = new SubGraphOrderer(mainGraphOrder, bottomGraph,
+        overlapThread);
+    overlapThread.start();
+    topGraphOrderer.start();
+    bottomGraphOrderer.start();
+
+    ArrayList<GraphNodeOrder> orderedTopGraph = topGraphOrderer.getNodeOrder();
+    ArrayList<GraphNodeOrder> orderedBottomGraph = bottomGraphOrderer.getNodeOrder();
+
+    //alignOverlappingNodes(orderedTopGraph, orderedBottomGraph);
+    removeEmptyLevels(orderedTopGraph, orderedBottomGraph);
+    // todo: recolapse bubble from the graphs
+    // todo: check if nodes are really overlapping (taking into account the bubbles)
+    return new Pair<>(orderedTopGraph, orderedBottomGraph);
+  }
+
+  /**
+   * Remove any level (graph depth level) which contains no nodes.
+   *
+   * @param topOrder    the order of the top graph.
+   * @param bottomOrder the order of the bottom graph.
+   */
+  private static void removeEmptyLevels(ArrayList<GraphNodeOrder> topOrder,
+      ArrayList<GraphNodeOrder> bottomOrder) {
+    int highestTopLevel = topOrder.get(topOrder.size() - 1).getLevel();
+    int highestBottomLevel = bottomOrder.get(bottomOrder.size() - 1).getLevel();
+    int highestLevel;
+    if (highestTopLevel > highestBottomLevel) {
+      highestLevel = highestTopLevel;
+    } else {
+      highestLevel = highestBottomLevel;
+    }
+    boolean[] levelIsNotEmpty = new boolean[highestLevel + 1];
+    for (GraphNodeOrder topNode : topOrder) {
+      levelIsNotEmpty[topNode.getLevel()] = true;
+    }
+    for (GraphNodeOrder bottomNode : bottomOrder) {
+      levelIsNotEmpty[bottomNode.getLevel()] = true;
+    }
+    fillEmptyLevels(topOrder, levelIsNotEmpty);
+    fillEmptyLevels(bottomOrder, levelIsNotEmpty);
+  }
+
+  /**
+   * Fill all of the empty graph levels by moving everything after it to the left.
+   *
+   * @param nodeOrder       the order of the nodes in the graph.
+   * @param levelIsNotEmpty an arraylist containing all of the graph levels which containg no nodes.
+   */
+  private static void fillEmptyLevels(ArrayList<GraphNodeOrder> nodeOrder,
+      boolean[] levelIsNotEmpty) {
+    int curLevel = 0;
+    int emptyLevels = 0;
+    for (GraphNodeOrder node : nodeOrder) {
+      if (node.getLevel() > curLevel) {
+        for (int i = curLevel; i < node.getLevel(); i++) {
+          if (!levelIsNotEmpty[i]) {
+            ++emptyLevels;
+          }
+        }
+        curLevel = node.getLevel();
+      }
+      node.addPositionOffset(-emptyLevels);
+    }
+  }
+
+  /**
+   * Thread which can be used to order a subgraph and mark the overlapping nodes.
+   */
+  private static class SubGraphOrderer extends Thread {
```

Comment:
I don't agree. This class is only used by the CompareSubgraphs class and is only used to execute a single method on a separate thead. If I wouldn't have multithreaded, this would simply have been 1 method, so in my opinion this class belongs within the CompareSubgraphs class.